bql/planner/data_access.go

```diff
@@ -409,62 +409,80 @@ func simpleFetch(ctx context.Context, gs []storage.Graph, cls *semantic.GraphCla
 // table. The semantic graph clause is also passed to be able to identify what
 // bindings to set.
 func addTriples(ts <-chan *triple.Triple, cls *semantic.GraphClause, tbl *table.Table) error {
-	for t := range ts {
+	shouldIgnoreTriple := func(t *triple.Triple) (bool, error) {
 		if cls.PID != "" {
 			// The triples need to be filtered.
 			if string(t.Predicate().ID()) != cls.PID {
-				continue
+				return true, nil
 			}
 			if cls.PTemporal {
 				if t.Predicate().Type() != predicate.Temporal {
-					continue
+					return true, nil
 				}
 				ta, err := t.Predicate().TimeAnchor()
 				if err != nil {
-					return fmt.Errorf("failed to retrieve time anchor from time predicate in triple %s with error %v", t, err)
+					return true, fmt.Errorf("failed to retrieve time anchor from time predicate in triple %s with error %v", t, err)
 				}
 				// Need to check the bounds of the triple.
 				if cls.PLowerBound != nil && cls.PLowerBound.After(*ta) {
-					continue
+					return true, nil
 				}
 				if cls.PUpperBound != nil && cls.PUpperBound.Before(*ta) {
-					continue
+					return true, nil
 				}
 			}
 		}
 		if cls.OID != "" {
 			if p, err := t.Object().Predicate(); err == nil {
 				// The triples need to be filtered.
 				if string(p.ID()) != cls.OID {
-					continue
+					return true, nil
 				}
 				if cls.OTemporal {
 					if p.Type() != predicate.Temporal {
-						continue
+						return true, nil
 					}
 					ta, err := p.TimeAnchor()
 					if err != nil {
-						return fmt.Errorf("failed to retrieve time anchor from time predicate in triple %s with error %v", t, err)
+						return true, fmt.Errorf("failed to retrieve time anchor from time predicate in triple %s with error %v", t, err)
 					}
 					// Need to check the bounds of the triple.
 					if cls.OLowerBound != nil && cls.OLowerBound.After(*ta) {
-						continue
+						return true, nil
 					}
 					if cls.OUpperBound != nil && cls.OUpperBound.Before(*ta) {
-						continue
+						return true, nil
 					}
 				}
 			}
 		}
+		return false, nil
+	}
+
+	var err error
+	for t := range ts {
+		ignoreTriple, err := shouldIgnoreTriple(t)
+		if err != nil {
+			break
+		}
+		if ignoreTriple {
+			continue
+		}
+
 		r, err := tripleToRow(t, cls)
 		if err != nil {
-			return err
+			break
 		}
 		if r != nil {
```

Comment:
super nit: in Go, it's more common to have the invalid/error logic nested. To make the code more standard you can always revert this logic.

if r == nil {
 // return error or error handling
}

// Normal logic -- adding row

Reply:
Thank you once again for the observation Jose.
I have just done so. 👍